instrumentation/apache-httpasyncclient-4.1/src/test/java/io/opentelemetry/instrumentation/hypertrace/apachehttpasyncclient/ApacheAsyncClientInstrumentationModuleTest.java

```diff
@@ -16,138 +16,88 @@
+import io.opentelemetry.javaagent.instrumentation.api.Pair;
 import java.io.BufferedReader;
-import java.io.ByteArrayInputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.InputStreamReader;
-import java.io.UnsupportedEncodingException;
 import java.net.URI;
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
-import java.util.List;
+import java.util.Map;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
-import java.util.concurrent.TimeoutException;
-import org.apache.http.Header;
-import org.apache.http.HttpEntity;
 import org.apache.http.HttpResponse;
 import org.apache.http.client.methods.HttpGet;
 import org.apache.http.client.methods.HttpPost;
 import org.apache.http.concurrent.FutureCallback;
-import org.apache.http.entity.ContentType;
 import org.apache.http.entity.StringEntity;
 import org.apache.http.impl.nio.client.CloseableHttpAsyncClient;
 import org.apache.http.impl.nio.client.HttpAsyncClients;
-import org.apache.http.message.BasicHeader;
-import org.hypertrace.agent.core.instrumentation.HypertraceSemanticAttributes;
-import org.hypertrace.agent.testing.AbstractInstrumenterTest;
-import org.hypertrace.agent.testing.TestHttpServer;
-import org.hypertrace.agent.testing.TestHttpServer.GetJsonHandler;
+import org.hypertrace.agent.testing.AbstractHttpClientTest;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeAll;
-import org.junit.jupiter.api.Test;
 
-class ApacheAsyncClientInstrumentationModuleTest extends AbstractInstrumenterTest {
-
-  private static final String JSON = "{\"id\":1,\"name\":\"John\"}";
-  private static final TestHttpServer testHttpServer = new TestHttpServer();
+class ApacheAsyncClientInstrumentationModuleTest extends AbstractHttpClientTest {
 
   private static final CloseableHttpAsyncClient client = HttpAsyncClients.createDefault();
 
+  public ApacheAsyncClientInstrumentationModuleTest() {
+    super(true);
+  }
+
   @BeforeAll
-  public static void startServer() throws Exception {
-    testHttpServer.start();
+  public static void startClient() throws Exception {
     client.start();
   }
 
   @AfterAll
-  public static void closeServer() throws Exception {
-    testHttpServer.close();
+  public static void closeClient() throws IOException {
+    client.close();
   }
 
-  @Test
-  public void getJson()
-      throws ExecutionException, InterruptedException, TimeoutException, IOException {
-    HttpGet getRequest =
-        new HttpGet(String.format("http://localhost:%s/get_json", testHttpServer.port()));
-    getRequest.addHeader("foo", "bar");
-    Future<HttpResponse> futureResponse = client.execute(getRequest, new NoopFutureCallback());
-
-    HttpResponse response = futureResponse.get();
-    Assertions.assertEquals(200, response.getStatusLine().getStatusCode());
+  @Override
+  public Pair<Integer, String> doPostRequest(
```

Comment:
Let's be more idiomatic and return our `Response` type.

Reply:
the Response type is different for different clients. So I used String and Integer

Reply:
that is correct, in the abstract test class we should define this class

```
public static class Response {
 String body;
 String responseCode;
}
```
It's more idiomatic and it does not feel right to depend in `io.opentelemetry.javaagent.instrumentation.api.Pair;` in our tests